sa/sa.go

```diff
@@ -1352,23 +1352,58 @@ func (ssa *SQLStorageAuthority) DeactivateAuthorization(ctx context.Context, id
 	if err != nil {
 		return err
 	}
-	table := authorizationTable
-	oldStatus := core.StatusValid
+
 	if existingPending(tx, id) {
-		table = pendingAuthorizationTable
-		oldStatus = core.StatusPending
+		authzObj, err := tx.Get(&pendingauthzModel{}, id)
+		if err != nil {
+			return Rollback(tx, err)
+		}
+		if authzObj == nil {
+			// InternalServerError because existingPending already told us it existed
+			return berrors.InternalServerError("failure retrieving pending authorization")
+		}
+		authz := authzObj.(*pendingauthzModel)
+		if authz.Status != core.StatusPending {
+			_ = Rollback(tx, nil)
+			return berrors.WrongAuthorizationStateError("authorization not pending")
+		}
+		result, err := tx.Delete(authzObj)
+		if err != nil {
+			return Rollback(tx, err)
+		}
+		if result != 1 {
+			_ = Rollback(tx, nil)
+			return berrors.InternalServerError("wrong number of rows deleted: expected 1, got %d", result)
+		}
+		authz.Status = core.StatusDeactivated
+		err = tx.Insert(&authzModel{authz.Authorization})
+		if err != nil {
+			return Rollback(tx, err)
+		}
+	} else {
+		authzObj, err := tx.Get(&authzModel{}, id)
```

Comment:
It seems like the old `UPDATE ... SET ... WHERE` logic should work fine in this case (authz is already finalized). Why go with the full-object read/write here?